tests/infer/mcmc/test_nuts.py

```diff
@@ -380,3 +380,31 @@ def model(data):
     samples = mcmc.get_samples()
     posterior = posterior_replay(model, samples, data, num_samples=num_samples)
     assert_equal(posterior["rate"].mean(0), true_rate, prec=0.3)
+
+
+def test_structured_mass():
+    def model(cov):
+        x = pyro.sample("x", dist.Normal(0, 1000))
+        y = pyro.sample("y", dist.Normal(0, 1000))
+        z = pyro.sample("z", dist.Normal(0, 1000))
+        xyz = torch.stack([x, y, z])
+        pyro.sample("obs", dist.MultivariateNormal(torch.zeros(3), cov), obs=xyz)
+
+    xy_cov = torch.tensor([[2., 1.], [1., 3.]])
+    z_var = torch.tensor([4.])
+    cov = torch.zeros(3, 3)
+    cov[:2, :2] = xy_cov
+    cov[2, 2] = z_var
+
+    # smoke tests
+    for dense_mass in [True, False]:
+        kernel = NUTS(model, jit_compile=True, ignore_jit_warnings=True, full_mass=dense_mass)
+        mcmc = MCMC(kernel, num_samples=1, warmup_steps=1)
+        mcmc.run(cov)
+        assert kernel.inverse_mass_matrix[("x", "y", "z")].dim() == 1 + int(dense_mass)
+
+    kernel = NUTS(model, jit_compile=True, ignore_jit_warnings=True, full_mass=[("x", "y")])
```

Comment:
Could we make this a little more complex and add a `w` variable that is full, so `full_mass=[("w",), ("x","y")]`? Then `w` will be full, `(x,y)` will be full, and `z` will be diagonal? This seems like the most general block diagonal structure.

Reply:
Sure, I'll add.